xapian-letor/ranker/listnet_ranker.cc

```diff
@@ -0,0 +1,256 @@
+
+    return prob;
+}
+
+// Equation (6) in paper Cao et al. "Learning to rank: from pairwise approach to listwise approach."
+static vector<double>
+calculateGradient(vector<FeatureVector> feature_vectors, prob_distrib_vector prob) {
+    LOGCALL_STATIC_VOID(API, "calculateGradient", feature_vectors | prob_distrib_vector);
+
+    vector<double> gradient(feature_vectors[0].get_fcount(),0);
+    int list_length = feature_vectors.size();
+
+    // Hold ground truth probability distribution
+    vector<double> prob_y(prob[0]);
+    // Hold prediction score probability distribution
+    vector<double> prob_z(prob[1]);
+
+    for (int i = 0; i < list_length; i++) {
+	vector<double> fvals = feature_vectors[i].get_fvals();
+
+	for (int k = 0; k < int(fvals.size()); k++){
+
+	    double first_term = - prob_y[i] * fvals[k];
+	    gradient[k] += first_term;
+
+	    double second_term = prob_z[i] * fvals[k];
+	    gradient[k] += second_term;
+
+	}
+    }
+
+    return gradient;
+}
+
+static void
+updateParameters(vector<double> & new_parameters, vector<double> gradient, double learning_rate) {
+    LOGCALL_STATIC_VOID(API, "updateParameters", new_parameters | gradient | learning_rate);
+	int num = new_parameters.size();
+
+    for (int i = 0; i < num; i++){
+	new_parameters[i] -= gradient[i] * learning_rate;
+    }
+
+}
+
+void
+ListNETRanker::train_model() {
+    LOGCALL_VOID(API, "ListNETRanker::train_model", NO_ARGS);
+
+    std::cout << "ListNET model begin to train..." << endl;
+
+    // get the training data
+    std::vector<Xapian::FeatureVector> fvv = get_traindata();
+    int fvv_len = fvv.size();
+
+    int feature_cnt = -1;
+    if (fvv_len != 0) {
+	feature_cnt = fvv[0].get_fcount();
+    }
+    else {
+	std::cout << "The training data is NULL!" << endl;
+	exit(1); // TODO: Throw exception
+    }
+
+    // initialize the parameters for neural network
+    vector<double> new_parameters(feature_cnt, 0.0);
+
+    // iterations
+    for (int iter_num = 1; iter_num < iterations; ++iter_num) {
+
+	// initialize Probability distributions of y and z
+	prob_distrib_vector prob = initializeProbability(fvv, new_parameters);
+	// compute gradient
+	vector<double> gradient = calculateGradient(fvv, prob);
+	// update parameters: w = w - gradient * learningRate
+	updateParameters(new_parameters, gradient, learning_rate);
+    }
+
+    parameters = new_parameters;
+
+}
+
+void
+ListNETRanker::save_model_to_file(const char* output_filename) {
+    LOGCALL_VOID(API, "ListNETRanker::save_model_to_file", output_filename);
+
+    ofstream parameters_file;
+    parameters_file.open(output_filename);
+
+    int parameters_size = parameters.size();
+
+    std::ostringstream oss;
+    oss.precision (std::numeric_limits<double>::digits10);
+
+    for (int i = 0; i < parameters_size; ++i) {
+	oss << parameters[i] << endl;
+    }
+
+    parameters_file << oss.str();
+    parameters_file.close();
+
+    cout<< "ListNET model saved to \"" << output_filename << "\"" << endl;
```

Comment:
Please get rid of all use of `iostream` in the library code - if you need these for debug, you can use the logging functionality in `debuglog.h`.